### ISSUE TITLE
more function for Collect() and CASE keyword enhancement

### ISSUE THREAD

**Issue by raycola** [emqx/kuiper#710]
Hi,
I am using kuiper as rule engine for a solution based on Edgex Foundry. I need to create a rule  that send some result data to an rest sink... I d like to create a rule with sql query to make difference between a property reading value "X" and a property reading value acquired ten minutes earlier. In my case the events occur always 10 minutes.
I found two solutions:

1) **USE TWO STREAMS:** I am thinking to use two streams and put into my rule. I think you can compare the last element and the first element in the window if the last element is 10 minutes later than the first. Pseudo `SQL: SELECT last.x - first.x FROM demo WHERE last.timestamp - first.timestamp is about 10 min GROUP BY SLIDINGWINDOW(mm, 10);`

2) **USE COLLECT()**: this is sql of rule:
`"sql": "SELECT  collect(PROP) as r,  ( collect(*)[1]->PROP - collect(*)[0]->PROP ) as diff FROM demo WHERE  meta(device) = \"DEVICE1\" AND PROP != nil GROUP BY  SLIDINGWINDOW(mi, 20)`
The first time the "out of index" exception occurs because there is only one element but next times works fine.
Here my log:
```
time="2021-02-05 12:55:39" level=info msg="sink result for rule rule1: [{\"error\":\"run Select error: out of index: 1 of 1\"}]" file="sinks/log_sink.go:16" rule=rule1
time="2021-02-05 12:56:09" level=info msg="sink result for rule rule1: [{\"diff\":0,\"r\":[102,102]}]" file="sinks/log_sink.go:16" rule=rule1
time="2021-02-05 12:56:39" level=info msg="sink result for rule rule1: [{\"diff\":1,\"r\":[102,103]}]" file="sinks/log_sink.go:16" rule=rule1
time="2021-02-05 12:57:09" level=info msg="sink result for rule rule1: [{\"diff\":1,\"r\":[103,104]}]" file="sinks/log_sink.go:16" rule=rule1
time="2021-02-05 12:57:39" level=info msg="sink result for rule rule1: [{\"diff\":6,\"r\":[104,110]}]" file="sinks/log_sink.go:16" rule=rule1
```

Honestly I prefer the second one. Two streams are expensive in term of resources.
Following enhancements could resolve my exception and many other cases:

1. SQL **CASE** expression allows you to evaluate a list of conditions  (like Transact-SQL)
2. **cardinality(anyarray)** function that returns the total number of elements in the array, or 0 if the array is empty (like PostgrestSQL)

What do you think about that enhancements?

Thank for advance
Ray

P.S.there is an error on help page: https://docs.emqx.io/en/kuiper/latest/sqls/built-in_functions.html#aggregate-functions because array starts from 0 not 1
_"Get the **first** element's column 'a' value within the current windows..."_

**Comment by ngjaying** [comment]
Thanks @raycola for the very detail explanation. I also prefer the second one and agree with the two enhancements. We will prioritize and process this request and correct the doc. Thanks again.

**Comment by raycola** [comment]
Thanks for your answer. 
I have another question. How can I avoid to submit the rest request when an error occurs:
```
time="2021-02-08 13:10:55" level=error msg="Operation project error: run Select error: out of index: 1 of 1" file="nodes/operations.go:121" rule=rule1
time="2021-02-08 13:10:55" level=info msg="sink result for rule rule1: [{\"error\":\"run Select error: out of index: 1 of 1\"}]" file="sinks/log_sink.go:16" rule=rule1

```

but I have my rest endpoint receives this body:
```
[
      {
        "error": "run Select error: out of index: 1 of 1"
      }
]
```
I supposed actions as items of a pipeline and when an action failed the pipeline would not continue.

Any suggest is appreciated
Thanks in advance

**Comment by ngjaying** [comment]
Currently, the errors are propogated through the whole pipeline. I think we can export a configuration to just print out the error and ignore in the pipeline so that the sink will only receive data.

**Comment by ngjaying** [comment]
Hi @raycola ,

The **cardinality** funciton and [sendError option](https://github.com/emqx/kuiper/blob/master/docs/en_US/rules/overview.md) to configure the propogation of error are implemented in 1.1.2 release.

The [**case expression**](https://github.com/emqx/kuiper/blob/master/docs/en_US/sqls/query_language_elements.md#case-expression) has merged to the repo and will be supported in the next release. 

Please check. Thanks.

**Comment by raycola** [comment]
Hi @ngjaying ,
the sendError option works well! No error was propagated as you can see. In my testplan I have an http exporter and the rule for diff and the events number is  correctly  an odd number (5) because the first rule's iteration was skipped.
File tracing  in stream.log:
`... level=error msg="Operation 5_project error: run Select error: out of index: 1 of 1" file="nodes/operations.go:122" rule=rule1`

my json_server endpoint:

```
{
  "events": [
    {
      "source": "edgex/app-services/http-export",
      "eventId": "38f9fa32-4f90-4e55-84a3-8c164a49a02f",
      "timestamp": "1615924531014457147",
      "updatings": [
        {
          "device": "mydevice",
          "readings": [
            {
              "name": "AI3",
              "value": "31",
              "valueType": "Number",
              "timestamp": "1615924531014378470"
            }
          ]
        }
      ],
      "id": 1
    },
    {
      "source": "kuiper",
      "eventId": "38f9fa32-4f90-4e55-84a3-8c164a49a02f",
      "timestamp": "1.615924531014457e+18",
      "updatings": [
        {
          "device": "mydevice",
          "readings": [
            {
              "name": "AI3_DIFF",
              "value": "14",
              "valueType": "Number",
              "timestamp": "1.6159245310143785e+18"
            }
          ]
        }
      ],
      "id": 2
    },
    {
      "source": "edgex/app-services/http-export",
      "eventId": "39f75e6d-f934-4e76-8e94-0fd2ca0ffa83",
      "timestamp": "1615924561018298721",
      "updatings": [
        {
          "device": "mydevice",
          "readings": [
            {
              "name": "AI3",
              "value": "45",
              "valueType": "Number",
              "timestamp": "1615924561018223332"
            }
          ]
        }
      ],
      "id": 3
    },
    {
      "source": "kuiper",
      "eventId": "39f75e6d-f934-4e76-8e94-0fd2ca0ffa83",
      "timestamp": "1.6159245610182986e+18",
      "updatings": [
        {
          "device": "mydevice",
          "readings": [
            {
              "name": "AI3_DIFF",
              "value": "23",
              "valueType": "Number",
              "timestamp": "1.6159245610182234e+18"
            }
          ]
        }
      ],
      "id": 4
    },
    {
      "source": "edgex/app-services/http-export",
      "eventId": "d44b8a64-01dc-4f97-a1eb-9cc15f8f4c79",
      "timestamp": "1615924591022233445",
      "updatings": [
        {
          "device": "mydevice",
          "readings": [
            {
              "name": "AI3",
              "value": "68",
              "valueType": "Number",
              "timestamp": "1615924591022138096"
            }
          ]
        }
      ],
      "id": 5
    }
  ]
}
```

I tested also cardinality function :
`"sql": "SELECT  cardinality(collect(*)) as number_of_items, .... "`
and it works well:
`time="2021-03-16 20:16:31" level=info msg="sink result for rule rule1: [{\"number_of_items\":2,....`

waiting for "CASE"  expression, this solution fulfills myrequirements 
Thank you very much
@raycola

**Comment by ngjaying** [comment]
@raycola Awesome. Thank you for verifying it.

**Comment by ngjaying** [comment]
The case expression is finally released in v1.2.0, so close. Thanks.
